DatadogCore/Sources/Core/Upload/DataUploadWorker.swift

```diff
@@ -62,56 +63,67 @@ internal class DataUploadWorker: DataUploadWorkerType {
             guard let self = self else {
                 return
             }
-
             let context = contextProvider.read()
             let blockersForUpload = self.uploadConditions.blockersForUpload(with: context)
             let isSystemReady = blockersForUpload.isEmpty
-            let nextBatch = isSystemReady ? self.fileReader.readNextBatch() : nil
-            if let batch = nextBatch {
-                self.backgroundTaskCoordinator?.beginBackgroundTask()
-                DD.logger.debug("⏳ (\(self.featureName)) Uploading batch...")
-
-                do {
-                    // Upload batch
-                    let uploadStatus = try self.dataUploader.upload(
-                        events: batch.events,
-                        context: context
-                    )
-
-                    // Delete or keep batch depending on the upload status
-                    if uploadStatus.needsRetry {
-                        self.delay.increase()
-
-                        DD.logger.debug("   → (\(self.featureName)) not delivered, will be retransmitted: \(uploadStatus.userDebugDescription)")
-                    } else {
-                        self.fileReader.markBatchAsRead(batch, reason: .intakeCode(responseCode: uploadStatus.responseCode ?? -1)) // -1 is unexpected here
-                        self.delay.decrease()
-
-                        DD.logger.debug("   → (\(self.featureName)) accepted, won't be retransmitted: \(uploadStatus.userDebugDescription)")
+            let files = isSystemReady ? self.fileReader.readFiles(maxBatchesPerUpload) : nil
+            var allUploadsSucceeded = false
+            if let files = files, files.isEmpty == false {
+                for file in files {
+                    self.backgroundTaskCoordinator?.beginBackgroundTask()
+                    guard let batch = self.fileReader.readBatch(from: file) else {
+                        continue
                     }
-
-                    switch uploadStatus.error {
-                    case .unauthorized:
-                        DD.logger.error("⚠️ Make sure that the provided token still exists and you're targeting the relevant Datadog site.")
-                    case let .httpError(statusCode: statusCode):
-                        telemetry.error("Data upload finished with status code: \(statusCode)")
-                    case let .networkError(error: error):
-                        telemetry.error("Data upload finished with error", error: error)
-                    case .none: break
+                    DD.logger.debug("⏳ (\(self.featureName)) Uploading batch...")
```

Comment:
Do we have any thoughts about these logs? It will become extremely spammy once we enable 10 uploads per track. // cc @maciejburda @maxep

Reply:
Yeah good point. I will revisit :)